File: Algorithms/DynamicProgramming/CpSolns/maxCircleSubarray-918.cpp
// Leetcode #918 - Maximum sum circular subarray

#include<iostream>
#include<vector>
#include<limits.h>

using std::vector;
using std::cout; using std::endl;
using std::min; using std::max;

class Solution{
    public:
        int maxSum(vector<int>& arr){
            int total = 0;
            int maxEnding = 0, minEnding = 0;
            int maxSum = INT_MIN, minSum = INT_MAX;
            for(int x : arr){
                total += x;
                maxEnding = max(maxEnding + x, x);
                minEnding = min(minEnding + x, x);
                maxSum = max(maxSum, maxEnding);
                minSum = min(minSum, minEnding);
            }
            if(total > 0) return max(maxSum, total - minSum);
            return maxSum;
        }
};

int main(){
    Solution soln;
    vector<int> n1 = {1, -2, 3, -2};
    vector<int> n2 = {5, -3, 5};
    cout << soln.maxSum(n1) << endl;
    cout << soln.maxSum(n2) << endl;
    return 0;
}

File: Algorithms/DynamicProgramming/JavaSolns/MaxCircleSubArray-918.java
// Leetcode #918

class MaxCircleSubArray{
    public static int maxSum(int[] nums){
        int maxSum = Integer.MIN_VALUE, minSum = Integer.MAX_VALUE;
        int maxEnding = 0, minEnding = 0, total = 0;
        
        for(int n : nums){
            total += n;
            maxEnding = Math.max(maxEnding + n, n);
            maxSum = Math.max(maxEnding, maxSum);
            minEnding = Math.min(minEnding + n, n);
            minSum = Math.min(minEnding, minSum);
        }
        return (maxSum > 0) ? Math.max(maxSum, total - minSum) : maxSum;
    }
    public static void main(String[] args){
        int[] n = {1,-2,3,-2};
        int[] n1 = {5,-3,5};
        int[] n2 = {-3,-2,-3};
        System.out.println(maxSum(n));
        System.out.println(maxSum(n1));
        System.out.println(maxSum(n2));
    }
}
